### PR TITLE
Fix exp_pauli issues on remote simulators and quantum devices

### PR DESCRIPTION
**Pull request by annagrin**
### Description
Fix exp_pauli issues on remote simulators and quantum devices

- Make DecompositionPatterns recognize code generated by argument synthesis for pauli word parameters
- Fix an issue in quake-to-qir pass where the IR produced had incorrect type for a store destination.

Closes: https://github.com/NVIDIA/cuda-quantum/issues/2222
Closes: https://github.com/NVIDIA/cuda-quantum/issues/2227

### CODE REVIEW

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp [hunk 1]
```diff
@@ -335,18 +335,52 @@ struct ExpPauliDecomposition : public OpRewritePattern<quake::ExpPauliOp> {
...
-    if (!defOp)
+    if (!optPauliWordStr.has_value()) {
       return failure();
+    }
+
+    auto pauliWordStr = optPauliWordStr.value();
+
+    // Remove optional last zero character
+    auto size = pauliWordStr.size();
+    if (pauliWordStr[size - 1] == '\0')
```
Comment (1tnguyen) [inline lib/Optimizer/Transforms/DecompositionPatterns.cpp:379]: nit: do we need to check if `size > 0` as well?

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

File: targettests/execution/exp_pauli.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// RUN: nvq++ %cpp_std --enable-mlir --target remote-mqpu          -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+//
+// Quantum emulators
+// RUN: nvq++ %cpp_std --target quantinuum               --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// RUN: nvq++ %cpp_std --target ionq                     --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// 2 different IQM machines for 2 different topologies
+// RUN: nvq++ %cpp_std --target iqm --iqm-machine Adonis --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// RUN: nvq++ %cpp_std --target iqm --iqm-machine Apollo --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// RUN: nvq++ %cpp_std --target oqc                      --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// clang-format on
```
Comment (khalatepradnya) [inline targettests/execution/exp_pauli.cpp:21]: For sake of completeness, I suggest adding `--target anyon` to this list.

File: targettests/execution/exp_pauli.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  counts = cudaq::sample(test_param, cudaq::pauli_word{"XY"});
+  printCounts(counts);
+  return 0;
+}
+
+// CHECK: 00
+// CHECK: 11
+
+// CHECK: 00
+// CHECK: 11
```
Comment (khalatepradnya) [inline targettests/execution/exp_pauli.cpp:61]: Nice to have: An end-to-end Python test similar to this C++ test. (Okay to implement on any one of the targets only.)

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 1 above)
Comment (annagrin) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: yes, thanks, done!

File: targettests/execution/exp_pauli.cpp (hunk 2 above)
Comment (annagrin) [reply targettests/execution/exp_pauli.cpp:22]: done, thanks!

File: targettests/execution/exp_pauli.cpp (hunk 3 above)
Comment (annagrin) [reply targettests/execution/exp_pauli.cpp:62]: Added a few tests for various backends.

File: targettests/execution/exp_pauli.cpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 // RUN: nvq++ %cpp_std --target iqm --iqm-machine Adonis --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
 // RUN: nvq++ %cpp_std --target iqm --iqm-machine Apollo --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
 // RUN: nvq++ %cpp_std --target oqc                      --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
+// RUN: nvq++ %cpp_std --target anion                    --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
```
Comment (khalatepradnya) [inline targettests/execution/exp_pauli.cpp:21]:
```suggestion
// RUN: nvq++ %cpp_std --target anyon                    --emulate -fkernel-exec-kind=2 %s -o %t && %t | FileCheck %s
```

Comment (khalatepradnya) [reply targettests/execution/exp_pauli.cpp:21]: nit: Also, other files have the targets listed in alphabetical order.

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp [hunk 5]
```diff
@@ -335,18 +335,52 @@ struct ExpPauliDecomposition : public OpRewritePattern<quake::ExpPauliOp> {
...
+            }
+          }
+        }
+      }
+    }
+
     // Assert that we have a constant known pauli word
-    auto defOp = pauliWord.getDefiningOp<cudaq::cc::CreateStringLiteralOp>();
-    if (!defOp)
+    if (!optPauliWordStr.has_value()) {
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/DecompositionPatterns.cpp:371]: nit: no braces needed here

**Review by schweitzpgi — APPROVED** [review]
LGTM

File: lib/Optimizer/Transforms/DecompositionPatterns.cpp (hunk 5 above)
Comment (annagrin) [reply lib/Optimizer/Transforms/DecompositionPatterns.cpp]: done! Created an issue to collect all the formatting requests that can be automated: https://github.com/NVIDIA/cuda-quantum/issues/2238

**Pull request merged**